compiler/optimizer/DeadTreesElimination.cpp

```diff
@@ -1041,7 +1041,35 @@ int32_t TR::DeadTreesElimination::process(TR::TreeTop *startTree, TR::TreeTop *e
                TR::Node *lastNode = lastTree->getNode();
                TR::Node *prevLastNode = prevLastTree->getNode();
 
-               if (lastNode->getOpCode().isIf() && !lastNode->getOpCode().isCompBranchOnly() &&
+               // If the node being moved is volatile, the order of the trees cannot
+               // be changed because that is the order containsNode used to check if
+               // canMoveIfVolatile should be true or not.
+               // For example, if the order of the trees look like below, adjusting
+               // the lastTree and prevLastTree in the following code would end up
+               // mistakenly reordering the volatile regStore nodes.
+               //
+               // E.g
+               // === Before adjusting lastTree & prevLastTree
+               // TT1
+               // TT2 (Volatile regStore) <--- prevLastTree
+               // TTBeingMoved (Volatile regStore)
+               // TT3 <--- lastTree
+               //
+               // === After adjusting lastTree & prevLastTree
+               // TT1 <--- prevLastTree
+               // TTBeingMoved (Volatile regStore) <--- Incorrect!!!
+               // TT2 (Volatile regStore) <--- lastTree
+               // TT3
+
+               bool skipMoveLastTreeIfVolatile = false;
+               if (iter.currentTree()->getNode()->getOpCode().isStoreReg() &&
```

Comment:
This is likely due to my unfamiliarity with this code, but is this phenomenon only an issue for regStores and not field stores?

Reply:
That's a good question. This change will only be run when the node (`isStoreReg` && RC > 1). If the node is not StoreReg, [the node is removed](https://github.com/eclipse/omr/blob/3cad1dcdcdc6867490787496e2b0a3efa439e7e1/compiler/optimizer/DeadTreesElimination.cpp#L985-L991). I think the check on `isStoreReg` is redundant. `node` here is `iter.currentTree()->getNode()`. I think I can also simplify the code to just check `node->getFirstChild()->mightHaveVolatileSymbolReference()` directly here. 

```
         if (!node->getOpCode().isStoreReg() || (node->getFirstChild()->getReferenceCount() == 1))
            {
            // Actually going to remove the treetop now
            //
            if (performTransformation(comp(), "%sRemove tree : [" POINTER_PRINTF_FORMAT "] ([" POINTER_PRINTF_FORMAT "] = %s)\n", optDetailString(), node, node->getFirstChild(), node->getFirstChild()->getOpCode().getName()))
               {
               ...
               }
            }
         else
            {
            if (performTransformation(comp(), "%sMove tree : [" POINTER_PRINTF_FORMAT "]([" POINTER_PRINTF_FORMAT "] = %s) to end of block\n", optDetailString(), node, node->getFirstChild(), node->getFirstChild()->getOpCode().getName()))
               {
               ...
               // This change 
               ... 
               }
            }
```

Reply:
@0xdaryl @vijaysun-omr updated in d08229b. Ready for another review